streamalert/alert_processor/outputs/teams.py

```diff
@@ -0,0 +1,217 @@
+"""
+Copyright 2017-present, Airbnb Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+from collections import OrderedDict
+
+import pymsteams
+from pymsteams import TeamsWebhookException
+
+from streamalert.alert_processor.helpers import compose_alert
+from streamalert.alert_processor.outputs.output_base import (
+    OutputDispatcher,
+    OutputProperty,
+    StreamAlertOutput,
+)
+from streamalert.shared.logger import get_logger
+
+LOGGER = get_logger(__name__)
+
+
+@StreamAlertOutput
+class TeamsOutput(OutputDispatcher):
+    """TeamsOutput handles all alert dispatching for Microsoft Teams"""
+
+    __service__ = "teams"
+
+    @classmethod
+    def get_user_defined_properties(cls):
+        """Get properties that must be assigned by the user when configuring a new Microsoft Teams
+        output.  This should be sensitive or unique information for this use-case that needs
+        to come from the user.
+
+        Every output should return a dict that contains a 'descriptor' with a description of the
+        integration being configured.
+
+        Microsoft Teams also requires a user provided 'webhook' url that is composed of the Team's
+        api url and the unique integration key for this output. This value should be should be
+        masked during input and is a credential requirement.
+
+        Returns:
+            OrderedDict: Contains various OutputProperty items
+        """
+        return OrderedDict(
+            [
+                (
+                    "descriptor",
+                    OutputProperty(
+                        description="a short and unique descriptor for this service configuration "
+                        "(ie: name of Team the webhook relates too)"
+                    ),
+                ),
+                (
+                    "url",
+                    OutputProperty(
+                        description="the full teams webhook url, including the secret",
+                        mask_input=True,
+                        cred_requirement=True,
+                    ),
+                ),
+            ]
+        )
+
+    @classmethod
+    def _format_message(cls, alert, publication, webhook_url):
+        """Format the message to be sent to Teams
+
+        Args:
+            alert (Alert): The alert
+            publication (dict): Alert relevant to the triggered rule
```

Comment:
`webhook_url` should be a string, right? Can you add the information about it here as well?

Reply:
Good spot, will add this to the `__docstring__'